test/io/event/selector/file_io.rb

```diff
@@ -29,32 +29,40 @@
 		let(:file) {Tempfile.new}
 		
 		it "can read using a buffer" do
+			write_result = nil
+			read_result = nil
+			
 			writer = Fiber.new do
 				buffer = IO::Buffer.new(128)
 				file.seek(0)
-				expect(selector.io_write(Fiber.current, file, buffer, 128)).to be == 128
```

Comment:
These assertions weren't running with URing selector. I think this is the reason:

https://github.com/socketry/io-event/blob/1e6e0e72e520e6d8c30e4c2089d89e741b3c6cba/ext/io/event/selector/uring.c#L454-L456

So, a sqe is is created and then:
- The `io_uring_submit_pending()` just increments `data->pending`: `data->pending += 1;`. This seems unusual, shouldn't this step actually perform submit?
- And then the next steps yields.

Effectively, we just create sqe without submitting it and then yield.

Reply:
This is a great question.

It turns out there is an overhead to submitting the SQEs. We try to batch as many as possible. We add the SQE and then increment pending count. When we enter the main event loop `select`, we submit them all at once.

The trade off is latency. The earlier we start the operation, the more likely it will be completed sooner. However, submitting all the operations 1 by 1 is a ton of system calls. I've experimented with submitting SQEs after a certain number accumulated. There are basically just different performance trade offs.

In terms of testing, it sounds like we might need to revisit some of the assumptions about what's done when a fiber is executed. It might just be a matter of calling `selector.select(0)` to flush the operations.

Reply:
This makes sense, thank you for the clarification. I'll try to make the specs work for uring with this insight.

Off-topic, what are your thoughts [about polling mode](https://unixism.net/loti/tutorial/sq_poll.html) to completely get rid of syscalls after setup? I assume the concern is high CPU? And "high IO amount" use-case is probably intended for C, not Ruby?

Reply:
I've tried something similar to polling mode. https://github.com/socketry/io-event/tree/uring-pthread-submission - because it happens on a background thread you have to use a mutex for synchronisation. I felt like the cost wasn't worth the benefit, at least not when I tried it. Worth exploring.